src/polydb.jl

```diff
@@ -0,0 +1,284 @@
+module Polydb
+
+import Polymake: call_function
+
+using Polymake
+
+using Mongoc
+
+#Polymake.Polydb's types store information via
+# a corresponding Mongoc type variable
+struct Database
+   mdb::Mongoc.Database
+end
+
+struct Collection
+   mcol::Mongoc.Collection
+end
+
+struct Cursor
+   mcursor::Mongoc.Cursor{Mongoc.Collection}
+end
+
+# alternative iterators returning `BSON` objects
+struct BSONCollection
+   mcol::Mongoc.Collection
+end
+
+struct BSONCursor
+   mcursor::Mongoc.Cursor{Mongoc.Collection}
+end
+
+# connects to the Polydb and
+# returns a Polymake.Polydb.Database instance
+function get_db()
+   client = Mongoc.Client("mongodb://polymake:database@db.polymake.org/?authSource=admin&ssl=true")
+   return Database(client["polydb"])
+end
+
+# returns a Polymake.Polydb.Collection instance with the given name
+# sections and collections in the name are connected with the '.' sign,
+# i.e. names = "Polytopes.Lattice.SmoothReflexive"
+function get_collection(db::Database, name::String)
+   return Collection(db.mdb[name])
+end
+
+# search a collection for documents matching the criteria given by d
+function find(c::Collection, d::Dict=Dict(); opts::Union{Nothing, Dict}=nothing)
+   return Cursor(Mongoc.find(c.mcol, Mongoc.BSON(d); options=opts))
+end
+
+function find(c::Collection, d::Pair...)
+   return Cursor(Mongoc.find(c.mcol, Mongoc.BSON(d...)))
+end
+
+# querying a `BSONCollection` returns a `BSONCursor`
+function find(c::BSONCollection, d::Dict=Dict(); opts::Union{Nothing, Dict}=nothing)
+   return BSONCursor(Mongoc.find(c.mcol, Mongoc.BSON(d); options=opts))
+end
+
+function find(c::BSONCollection, d::Pair...)
+   return BSONCursor(Mongoc.find(c.mcol, Mongoc.BSON(d...)))
+end
+
+# creating `BSON` iterators from the respective `Polymake.BigObject` iterator
+function BSONCollection(c::Collection)
+   return BSONCollection(c.mcol)
+end
+
+function BSONCursor(cursor::Cursor)
+   return BSONCursor(cursor.mcursor)
+end
+
+# returns a Polymake.BigObject from a Mongoc.BSON document
+function parse_document(bson::Mongoc.BSON)
+   str = Mongoc.as_json(bson)
+   return call_function(:common, :deserialize_json_string, str)
+end
+
+#Iterator
+
+# help functions to reduce runtime and keep code clean
+function _evaluateIteration(nothing::Nothing)
```

Comment:
did you have a look at how `Mongoc` handles iteration?

It's only cursor that needs non-intuitive treatment:
```julia
function _iterate(cursor::Cursor, state::Nothing=nothing)
    bson_handle_ref = Ref{Ptr{Cvoid}}()
    has_next = mongoc_cursor_next(cursor.handle, bson_handle_ref)

    if has_next
        # The bson document is valid only until the next call to mongoc_cursor_next.
        # So we should return a deepcopy.
        return deepcopy(BSON(bson_handle_ref[], enable_finalizer=false)), nothing
    else
        err_ref = Ref{BSONError}()
        if mongoc_cursor_error(cursor.handle, err_ref)
            throw(err_ref[])
        end

        return nothing
    end
end

Base.iterate(cursor::Cursor, state::Nothing=nothing) = _iterate(cursor, state)

function Base.iterate(coll::Collection)
    cursor = find(coll)
    return iterate(coll, cursor)
end

function Base.iterate(coll::Collection, state::Cursor)
    next = _iterate(state)
    if next == nothing
        return nothing
    else
        doc, _ = next
        return doc, state
    end
end
```
since our `Cursor/Collection` tightly wrap `Mongoc`s, can we copy it here?

Reply:
this implements iteration over `Polydb.Cursor`: (nothing else is needed)
```julia
function Base.iterate(cursor::Polymake.Polydb.Cursor{Polymake.BigObject}, state::Nothing=nothing)
    next = iterate(cursor.mcursor, state)
    isnothing(next) && return nothing
    return Polymake.Polydb.parse_document(first(next)), nothing # or last(next)
end

Base.iterate(cursor::Polymake.Polydb.Cursor{Mongoc.BSON}, state::Nothing=nothing) = 
    iterate(cursor.mcursor, state)
```
it seems that (a literal copy of their `iterate`s):
```julia
Base.iterate(coll::Polymake.Polydb.Collection{T}) where T = 
    return iterate(coll, Cursor{T}(find(coll)))

function Base.iterate(coll::Polymake.Polydb.Collection, state::Polymake.Polydb.Cursor)
    next = iterate(state, nothing)
    isnothing(next) && return nothing
    doc, _ = next
    return doc, state
end 
```
should work for collections. small unions are pretty well optimized so there is no need to worry about them (if this was your intention for defining `_evaluateIteration`).